### DISCUSSION TITLE
CGNat and Hole Punching

### DISCUSSION THREAD

**Discussion by AaronLS** [StrangeLoopGames/EcoIssues#22873] [Q&A] [answered]
I've seen UPnP and port "punch-through" mentioned in server setup documentation.

I am behind Carrier Grade NAT, which means UPnP will not work and port forwarding is not an option.  However hole punching often does work.

Typically true hole-punching involves a third party coordinating the initial connection.  I am assuming the SLG service that maintains the list of public servers acts as that coordinator?  I.e. when a client attempts to connect to a server, then it first connect to SLG service which coordinates the hole-punching?

Just trying to confirm that you're doing true hole punching before a I make a purchase.  I understand it's not 100% reliable since it depends on some of the predictable-ness of port assignments on the NAT, but it's the only way I am able to host successfully without shelling out for VPN's or static IP.  

Thanks in advance.

**Comment by mirasrael** [comment, marked as answer]
Hi. We have NAT punch-through solution which works throw Master Server as you assumed. But it doesn't work in some cases like symmetric NAT. Also there Relay server which covers cases when neither UPnP nor Punch Through works, but it then will relay all traffic to the server and from the server, so you may expect higher ping and it isn't recommended to use outside of scope of few friends multiplayer.
If you're going to host a server then it is recommended to have direct IP anyway, but you can actually test if it works without purchase. Eco Server may be started by everyone without purchase. If you aware if you can connect with client then it depends on a server - if server has public IP then you don't need to have dedicated IP or UPnP on your side.
